### ISSUE TITLE
dictionary changed size during iteration

### ISSUE THREAD

**Issue by AndreLouisCaron** [cherrypy/cheroot#317]
Hi there!

I've just upgraded from `cheroot==8.4.4` to `cheroot==8.4.5` and I noticed what looks like a regression.

❓ **I'm submitting a ...**
- [X] 🐞 bug report

🐞 **Describe the bug. What is the current behavior?**

I see this in my server logs:

```
Error in HTTPServer.tick
Traceback (most recent call last):
  File ".../lib/site-packages/cheroot/server.py", line 1798, in serve
    self.tick()
  File ".../lib/site-packages/cheroot/server.py", line 2046, in tick
    self._connections.expire()
  File ".../lib/site-packages/cheroot/connections.py", line 105, in expire
    in self._selector.get_map().items()
  File ".../lib/_abcoll.py", line 414, in items
    return [(key, self[key]) for key in self]
RuntimeError: dictionary changed size during iteration
```

❓ **What is the motivation / use case for changing the behavior?**

There is occasional log spam (for a low traffic environment), but I assume this has some other undesired side-effects...

💡 **To Reproduce**

I'm not sure exactly.  We have a test environment with synthetic checks that does a couple requests per second and the behaviour occurs "naturally"...

📋 **Details**

We very recently upgraded from version 8.4.4 to 8.4.5, but it's possible the error was already present in 8.4.4.  We use `pip-compile` and `cheroot` is not pinned, so we usually get updates shortly after they are released.

📋 **Environment**
- Cheroot version: 8.4.5
- CherryPy version: 17.4.2 (I don't think we really use it)
- Python version: 2.7
- OS: Windows
- Browser: requests

**Comment by cyraxjoe** [comment]
:disappointed: I think that this issue is bound to happen unless we use a lock or we _really_ have a perfect management of the underlying selectors. But given how I saw that the `selectors` module implemented the mapping we have to manage that contention ourselves.

This situation could have happen if concurrently a [worker thread is putting a connection in a thread](https://github.com/cherrypy/cheroot/blob/master/cheroot/workers/threadpool.py#L123) and the [server is expiring connections.](https://github.com/cherrypy/cheroot/blob/master/cheroot/server.py#L2046)

1. WorkerThread -> returning a connection to be managed because is using `keep-alive`.
2. ServerThread -> expiring connections, exactly at the time when is iterating the mapping.

The [alternative fix that I proposed](https://github.com/cyraxjoe/cheroot/commit/518eacece4db8092f34d54109c72d5af94fd957c) managed an explicit external set and used a lock to register/unregistered connections. The downside is that most definitely would have additional overhead. Probably not a whole lot, but at least we would have certainty that we weren't able to trigger this type of error, a race condition.

**Comment by cyraxjoe** [comment]
Even my _alternative_ fix is still prone to a race condition, because I'm not sure that `set.copy` is thread safe. I think not, probably very fast, but no guarantees. The implementation would need to lock while the copy instruction is been executed before returning the iterator.

**Comment by AndreLouisCaron** [comment]
FWIW, I have been using `cheroot` in production for a looooon time (multiple years) under very high load and it's the first time I see this type of error, so I would assume some recent change to connection management has accidentally introduced this...

> I'm not sure that set.copy is thread safe. I think not, probably very fast, but no guarantees

I'm not sure this applies to all interpreters, but CPython implements sets in C (lookup [setobject.c](https://github.com/python/cpython/blob/master/Objects/setobject.c)) and none of its operations release the GIL, so `set.copy` can be considered atomic/threadsafe.

**Comment by The-Compiler** [comment]
I also saw this (or at least a closely related bug) today, with cheroot 8.4.5. The end of the stacktrace looks slightly different:

```
Error in HTTPServer.tick
Traceback (most recent call last):
  File "/__w/qutebrowser/qutebrowser/.tox/py38/lib/python3.8/site-packages/cheroot/server.py", line 1798, in serve
    self.tick()
  File "/__w/qutebrowser/qutebrowser/.tox/py38/lib/python3.8/site-packages/cheroot/server.py", line 2046, in tick
    self._connections.expire()
  File "/__w/qutebrowser/qutebrowser/.tox/py38/lib/python3.8/site-packages/cheroot/connections.py", line 102, in expire
    timed_out_connections = [
  File "/__w/qutebrowser/qutebrowser/.tox/py38/lib/python3.8/site-packages/cheroot/connections.py", line 102, in <listcomp>
    timed_out_connections = [
  File "/usr/lib/python3.8/_collections_abc.py", line 743, in __iter__
    for key in self._mapping:
RuntimeError: dictionary changed size during iteration
```

**Comment by jaraco** [comment]
Fix from #337 is now released as 8.4.8. Please verify the fix and report back here. Thanks!

**Comment by AndreLouisCaron** [comment]
Thanks a lot for the fix!  I'm rolling that out to our test environments and will confirm the fix later this week :-)
